doc/bm.1
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "BM" "1" "" "bm-haskell 0.3.0.0 (2026-05-24)" "bm Manual"
.nh
.SH NAME
.PP
\f[V]bm\f[R] - open bookmarks and queries from the command line
.SH SYNOPSIS
.PP
\f[V]bm\f[R] [\f[I]OPTIONS\f[R]] [ARG \&...]
.SH DESCRIPTION
.PP
\f[V]bm\f[R] is a utility for opening bookmarks and queries from the
command line.
.SH OPTIONS
.TP
-h, --help
show help and exit
.TP
--version
show version and exit
.TP
-c, --config \f[I]CONFIG\f[R]
config file (default: \f[V]$XDG_CONFIG_HOME/bm.yaml\f[R])
.TP
-t, --trace
show trace output for debugging
.SH ARGUMENTS
.TP
\f[I]ARG\f[R]
keyword (prefix) to traverse the bookmark hierarchy
.SH EXIT CODES
.TP
0
no error
.TP
1
program error
.TP
2
usage error
.SH CONFIGURATION
.PP
Bookmarks and queries are configured in a YAML file.
The configuration file \f[V]$XDG_CONFIG_HOME/bm.yaml\f[R] is used by
default.
On non-Windows systems, this is usually \f[V]\[ti]/.config/bm.yaml\f[R].
.PP
The command used to open bookmarks and queries defaults to the following
commands, depending on the operating system:
.IP \[bu] 2
Linux: \f[V]xdg-open\f[R]
.IP \[bu] 2
Windows: \f[V]start\f[R]
.IP \[bu] 2
macOS: \f[V]open\f[R]
.PP
The configuration file has the following properties:
.TP
\f[I]command\f[R]
command used to open bookmarks and queries (string, optional)
.TP
\f[I]args\f[R]
list of bookmarks (array of bookmark objects)
.PP
If you specify a \f[V]command\f[R] here, it takes precedence over the
default described above.
.PP
Bookmark objects have the following properties:
.TP
\f[I]keyword\f[R]
keyword to match against command-line arguments (string)
.TP
\f[I]command\f[R]
command used to open bookmarks and queries (string, optional)
.TP
\f[I]url\f[R]
bookmark URL (string, optional)
.TP
\f[I]query\f[R]
query definition (query object, optional)
.TP
\f[I]args\f[R]
list of child bookmarks (array of bookmark objects)
.PP
Any scalar value can be used to define a \f[V]keyword\f[R].
If you specify a \f[V]command\f[R] for a bookmark, that command takes
precedence for that bookmark as well as any child bookmarks.
Only one of \f[V]query\f[R] and \f[V]args\f[R] may be specified.
.PP
Query objects have the following properties:
.TP
\f[I]action\f[R]
query action URL (string)
.TP
\f[I]parameter\f[R]
query parameter name (string, optional, default: \[lq]q\[rq])
.TP
\f[I]hidden\f[R]
list of hidden parameters (array of parameter object, optional)
.PP
Parameter objects have the following properties:
.TP
\f[I]name\f[R]
parameter name (string)
.TP
\f[I]value\f[R]
parameter value (string)
.PP
Any scalar value can be used to define a \f[V]value\f[R].
.SH COMPLETION
.PP
Run the following command to enable Bash completion:
.IP
.nf
\f[C]
$ source <(bm --complete-bash \[ga]which bm\[ga])
\f[R]
.fi
.SH TRACING
.PP
The \f[V]--trace\f[R] option shows trace output for debugging.
Commands are displayed in brackets such as \f[V][xdg-open]\f[R].
Bookmark keywords are displayed in angle brackets such as
\f[V]<keyword>\f[R].
When there is no error, the command that is executed to open the
bookmark/query is displayed without escaping.
.SH PROJECT
.TP
GitHub:
<https://github.com/ExtremaIS/bm-haskell>
.TP
Reporting issues:
GitHub: <https://github.com/ExtremaIS/bm-haskell/issues>
.RS
.PP
Email: <bugs@extrema.is>
.RE
.TP
Copyright
Copyright (c) 2021-2026 Travis Cardwell
.TP
License
The MIT License <https://opensource.org/licenses/MIT>


CHANGELOG.md
# `bm-haskell` Changelog

This project follows the [Haskell package versioning policy][PVP], with
versions in `A.B.C.D` format.  `A` may be incremented arbitrarily for
non-technical reasons, but [semantic versioning][SemVer] is otherwise
followed, where `A.B` is the major version, `C` is the minor version, and `D`
is the patch version.  Initial development uses versions `0.0.0.D`, for which
every version is considered breaking.

[PVP]: <https://pvp.haskell.org/>
[SemVer]: <https://semver.org/>

The format of this changelog is based on [Keep a Changelog][KaC], with the
following conventions:

* Level-two heading `Unreleased` is used to track changes that have not been
  released.
* Other level-two headings specify the release in `A.B.C.D (YYYY-MM-DD)`
  format, with newer versions above older versions.
* Level-three headings are used to categorize changes as follows:
    1. Breaking
    2. Non-Breaking
* Changes are listed in arbitrary order and present tense.

[KaC]: <https://keepachangelog.com/en/1.0.0/>

## Unreleased

## 0.3.0.0 (2026-05-24)

### Breaking

* Remove support for GHC 8.6, constraining lower bounds
* Change minimal Cabal from 1.24 to 3.0

### Non-Breaking

* Bump `aeson` dependency version upper bound
* Bump `base` dependency version upper bound
* Bump `filepath` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound
* Bump `tasty` dependency version upper bound
* Bump `text` dependency version upper bound

## 0.2.0.0 (2023-05-28)

### Breaking

* Add support for `optparse-applicative` `0.18`

### Non-Breaking

* Bump `ansi-wl-pprint` dependency version upper bound

## 0.1.1.0 (2023-04-23)

### Non-Breaking

* Bump `aeson` dependency version upper bound
* Bump `transformers` dependency version upper bound
* Bump `vector` dependency version upper bound
* Adjust dependency constraints to match tested versions

## 0.1.0.2 (2022-03-02)

### Non-Breaking

* Bump `text` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound

## 0.1.0.1 (2021-10-10)

### Non-Breaking

* Add dependency bounds

## 0.1.0.0 (2021-07-02)

### Breaking

* Initial release


src/BM.hs
------------------------------------------------------------------------------
-- |
-- Module      : BM
-- Description : API
-- Copyright   : Copyright (c) 2021-2026 Travis Cardwell
-- License     : MIT
------------------------------------------------------------------------------

{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeApplications #-}

module BM
  ( -- * Constants
    version
    -- * Types
  , Argument
  , Command
  , Error
  , Keyword
  , ParameterName
  , ParameterValue
  , Trace
  , Url
  , Config(..)
  , Bookmark(..)
  , Query(..)
  , Parameter(..)
  , Proc(..)
    -- * API
  , run
  , getCompletion
  ) where

-- https://hackage.haskell.org/package/aeson
import qualified Data.Aeson as A
import Data.Aeson (FromJSON, (.:), (.:?), (.!=))
import qualified Data.Aeson.Types as AT

-- https://hackage.haskell.org/package/base
import Data.List (intercalate, isPrefixOf)
import Data.Maybe (fromMaybe)
import Data.Version (showVersion)
import qualified System.Info

-- https://hackage.haskell.org/package/dlist
import qualified Data.DList as DList
import Data.DList (DList)

-- https://hackage.haskell.org/package/network-uri
import qualified Network.URI as URI

-- https://hackage.haskell.org/package/scientific
import qualified Data.Scientific as Sci

-- https://hackage.haskell.org/package/text
import qualified Data.Text as T

-- https://hackage.haskell.org/package/transformers
import Control.Monad.Trans.Writer (Writer, runWriter, tell)

-- https://hackage.haskell.org/package/vector
import qualified Data.Vector as V
import Data.Vector (Vector)

-- (bm:cabal)
import qualified Paths_bm as Project

------------------------------------------------------------------------------
-- $Constants

-- | bm version string (\"@bm-haskell X.X.X.X@\")
--
-- @since 0.1.0.0
version :: String
version = "bm-haskell " ++ showVersion Project.version

------------------------------------------------------------------------------

-- | Default command, depending on the OS
defaultCommand :: Command
defaultCommand = case System.Info.os of
    "mingw32" -> "start"
    "darwin"  -> "open"
    _other    -> "xdg-open"

-- | Default query parameter name
defaultParameter :: ParameterName
defaultParameter = "q"

------------------------------------------------------------------------------
-- $Types
--
-- This implementation makes heavy use of the 'String' type.  Type aliases are
-- provided to make the API easier to read.

-- | CLI argument or process argument
--
-- @since 0.1.0.0
type Argument = String

-- | Process command
--
-- This command is executed with a single URL argument to open a
-- bookmark/query.
--
-- @since 0.1.0.0
type Command = FilePath

-- | Error message
--
-- @since 0.1.0.0
type Error = String

-- | Bookmark keyword
--
-- The configuration file defines a hierarchy of keywords that are matched
-- against CLI arguments to determine which bookmark/query to open.
--
-- @since 0.1.0.0
type Keyword = String

-- | Query parameter name
--
-- @since 0.1.0.0
type ParameterName = String

-- | Query parameter value
--
-- @since 0.1.0.0
type ParameterValue = String

-- | Trace line for debugging
--
-- @since 0.1.0.0
type Trace = String

-- | Bookmark or query action URL
--
-- @since 0.1.0.0
type Url = String

------------------------------------------------------------------------------

-- | Configuration
--
-- YAML attributes:
--
-- * @command@: top-level command (string, default depends on the OS)
-- * @args@: bookmarks (array of t'Bookmark')
--
-- Default commands:
--
-- * Linux: @xdg-open@
-- * Windows: @start@
-- * macOS: @open@
--
-- @since 0.1.0.0
data Config
  = Config
    { configCommand :: !Command
    , configArgs    :: !(Vector Bookmark)
    }
  deriving Show

instance FromJSON Config where
  parseJSON = A.withObject "Config" $ \o ->
    Config
      <$> o .:? "command" .!= defaultCommand
      <*> o .:  "args"

------------------------------------------------------------------------------

-- | Bookmark definition
--
-- YAML attributes:
--
-- * @keyword@: bookmark keyword (string)
-- * @command@: command for this bookmark and children (string, optional)
-- * @url@: bookmark URL (string, optional)
-- * @query@: bookmark query definition (t'Query', optional)
-- * @args@: child bookmarks (array of t'Bookmark', optional)
--
-- A command be set to override the top-level command, but this is generally
-- not done.  If a bookmark is selected and there is no URL, the first child
-- is processed.  Only one of @query@ and @args@ may be present.
--
-- @since 0.1.0.0
data Bookmark
  = Bookmark
    { keyword     :: !Keyword
    , mCommand    :: !(Maybe Command)
    , mUrl        :: !(Maybe Url)
    , queryOrArgs :: !(Either Query (Vector Bookmark))
    }
  deriving Show

instance FromJSON Bookmark where
  parseJSON = A.withObject "Bookmark" $ \o -> do
    keyword  <- parseToString =<< o .: "keyword"
    mCommand <- o .:? "command"
    mUrl     <- o .:? "url"
    mQuery   <- o .:? "query"
    mArgs    <- o .:? "args"
    queryOrArgs <- case (mQuery, mArgs) of
      (Nothing,    Just args) -> pure $ Right args
      (Just query, Nothing)   -> pure $ Left query
      (Nothing,    Nothing)   -> pure $ Right V.empty
      (Just{},     Just{})    -> fail $
        "bookmark has both query and args: " ++ keyword
    pure Bookmark{..}

------------------------------------------------------------------------------

-- | Query definition
--
-- YAML attributes:
--
-- * @action@: URL (string)
-- * @parameter@: query parameter name (string, default: @q@)
-- * @hidden@: array of constant parameters (t'Parameter')
--
-- @since 0.1.0.0
data Query
  = Query
    { action           :: !Url
    , parameter        :: !ParameterName
    , hiddenParameters :: !(Vector Parameter)
    }
  deriving Show

instance FromJSON Query where
  parseJSON = A.withObject "Query" $ \o ->
    Query
      <$> o .:  "action"
      <*> o .:? "parameter" .!= defaultParameter
      <*> o .:? "hidden"    .!= V.empty

------------------------------------------------------------------------------

-- | HTTP GET parameter definition
--
-- YAML attributes:
--
-- * @name@: parameter name
-- * @value@: constant parameter value
--
-- @since 0.1.0.0
data Parameter
  = Parameter
    { name  :: !ParameterName
    , value :: !ParameterValue
    }
  deriving Show

instance FromJSON Parameter where
  parseJSON = A.withObject "Parameter" $ \o ->
    Parameter
      <$> o .: "name"
      <*> (parseToString =<< o .: "value")

-- | Encode an HTTP GET parameter
--
-- Spaces are transformed to plus characters, and other reserved characters
-- are escaped.
encodeParameter :: Parameter -> String
encodeParameter Parameter{..} = encodePart name ++ "=" ++ encodePart value
  where
    encodePart :: String -> String
    encodePart
      = map (\c -> if c == ' ' then '+' else c)
      . URI.escapeURIString ((||) <$> URI.isUnreserved <*> (== ' '))

------------------------------------------------------------------------------

-- | Process specification
--
-- @since 0.1.0.0
data Proc
  = Proc
    { command   :: !Command
    , arguments :: ![Argument]
    }
  deriving (Eq, Show)

------------------------------------------------------------------------------
-- $API

-- | Determine the process to execute for the given config and CLI arguments
--
-- @since 0.1.0.0
run
  :: Config
  -> [Argument]
  -> (Either Error Proc, [Trace])
run Config{..} cliArgs = fmap DList.toList . runWriter $ do
    trace $ formatCommand configCommand
    loop configCommand configArgs cliArgs
  where
    loop
      :: Command
      -> Vector Bookmark
      -> [Argument]
      -> Writer (DList Trace) (Either Error Proc)
    loop cmd bms (arg:args) = case V.find (isPrefixOf arg . keyword) bms of
      Just bm -> do
        trace $ formatBookmark bm
        case queryOrArgs bm of
          Left query
            | null args -> case mUrl bm of
                Just url -> openUrl (fromMaybe cmd $ mCommand bm) url
                Nothing -> returnError $ "no query for " ++ keyword bm
            | otherwise -> openQuery (fromMaybe cmd $ mCommand bm) query args
          Right bms'
            | null args -> case mUrl bm of
                Just url -> openUrl (fromMaybe cmd $ mCommand bm) url
                Nothing  -> case bms' V.!? 0 of
                  Just bm' ->
                    loop (fromMaybe cmd $ mCommand bm) bms' [keyword bm']
                  Nothing -> returnError $ "no URL for " ++ keyword bm
            | otherwise -> loop (fromMaybe cmd $ mCommand bm) bms' args
      Nothing -> returnError $ "unknown argument: " ++ arg
    loop _cmd _bms [] = returnError "no arguments"

    returnError :: Error -> Writer (DList Trace) (Either Error Proc)
    returnError = return . Left

    openUrl :: Command -> Url -> Writer (DList Trace) (Either Error Proc)
    openUrl cmd url = do
      trace $ unwords [cmd, url]
      return . Right $ Proc cmd [url]

    openQuery
      :: Command
      -> Query
      -> [Argument]
      -> Writer (DList Trace) (Either Error Proc)
    openQuery cmd Query{..} args
      = openUrl cmd
      . (action ++)
      . ('?' :)
      . intercalate "&"
      . map encodeParameter
      $ Parameter parameter (unwords args) : V.toList hiddenParameters

    trace :: Trace -> Writer (DList Trace) ()
    trace = tell . DList.singleton

    formatCommand :: Command -> Trace
    formatCommand = ('[' :) . (++ "]")

    formatKeyword :: Keyword -> Trace
    formatKeyword = ('<' :) . (++ ">")

    formatBookmark :: Bookmark -> Trace
    formatBookmark Bookmark{..} = case mCommand of
      Just command -> unwords [formatKeyword keyword, formatCommand command]
      Nothing      -> formatKeyword keyword

------------------------------------------------------------------------------

-- | Get CLI completion options
--
-- @since 0.1.0.0
getCompletion
  :: Config
  -> [Argument]  -- ^ current CLI arguments, last one being completed
  -> [Argument]  -- ^ completion options
getCompletion Config{..} = loop configArgs
  where
    loop :: Vector Bookmark -> [Argument] -> [Argument]
    loop bms [arg] = filter (isPrefixOf arg) . map keyword $ V.toList bms
    loop bms (arg:args) = case V.find (isPrefixOf arg . keyword) bms of
      Just bm -> case queryOrArgs bm of
        Left{} -> []
        Right bms' -> loop bms' args
      Nothing -> []
    loop _bms [] = []

------------------------------------------------------------------------------
-- $Internal

-- | Parse any scalar value as a string
--
-- Strings, numbers, booleans, and null are parsed as a string.  Arrays and
-- objects result in an error.
parseToString :: A.Value -> AT.Parser String
parseToString = \case
    (A.String t)  -> pure $ T.unpack t
    (A.Number n)  -> pure . either (show @Double) (show @Integer) $
      Sci.floatingOrInteger n
    (A.Bool b)    -> pure $ if b then "true" else "false"
    A.Null        -> pure "null"
    A.Array{}     -> fail "unexpected array"
    A.Object{}    -> fail "unexpected object"
